File: mixin-padang-garut/src/main/python/garut/functions/dataset/distinct_rows.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
import pandas as pd

from typing import Dict, List
from garut.contexts.context import Context
from garut.instruction import InstructionFactory
from garut.functions.function import FunctionRegistry
from garut.functions.dataset.group_rows import GroupRows
from garut.functions.dataset.dataset_function import DatasetFunction


class DistinctRows(DatasetFunction):

    FUNCTION_NAME = "DistinctRows"

    def execute(self, context: Context, options: Dict[str, any]) -> any:
        dataset: pd.DataFrame = self.getDataset(options)
        keys = options[GroupRows.KEYS]
        groupby = dataset.groupby(keys)
        result = groupby.size().reset_index()[keys]
        result = self.forceObjectToString(result)
        return result


parameters = [DatasetFunction.DATASET, GroupRows.KEYS]
registry: FunctionRegistry = FunctionRegistry.getInstance()
registry.register(DistinctRows.FUNCTION_NAME, DistinctRows, parameters)

factory: InstructionFactory = InstructionFactory.getInstance()
factory.registerInstruction(DistinctRows.FUNCTION_NAME)

File: mixin-padang-webface/src/main/js/padang/functions/dataset/DistinctRows.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import FunctionPlan from "padang/plan/FunctionPlan";
import FunctionPlanRegistry from "padang/plan/FunctionPlanRegistry";

import UniqueRows from "padang/functions/dataset/UniqueRows";

export default class DistinctRows extends UniqueRows {

    public static FUNCTION_NAME = "DistinctRows";

    constructor(
        public dataset: string,
        public keys: any[]) {
        super(DistinctRows.FUNCTION_NAME, dataset);
    }

    public static getPlan(): FunctionPlan {
        let plan = new FunctionPlan(
            DistinctRows.FUNCTION_NAME,
            "Distinct Rows",
            "mdi-format-list-bulleted-type",
            "Distinct rows based on keys"
        );
        let parameters = plan.getParameterList();
        parameters.add(DistinctRows.DATASET_PLAN);
        parameters.add(DistinctRows.KEYS_PLAN);
        return plan;
    }

}

let registry = FunctionPlanRegistry.getInstance();
registry.registerPlan(DistinctRows.getPlan());